File: frontend/src/api/ai.ts
import { apiClient } from './client';
import { Approval, AgentTelemetry } from '../types/ai';

export interface ChatResponsePayload {
  run_id: string;
  conversation_id: string;
  status: string;
  plan?: string;
  response: string;
  tool_calls_count: number;
  pending_approvals: Array<{
    approval_id: string;
    risk_level: 'LOW' | 'MEDIUM' | 'HIGH';
    action: string;
    payload: Record<string, any>;
    message: string;
  }>;
  tokens_used: number;
  execution_time_ms: number;
}

export interface WorkspaceAISettingsResponse {
  workspace_id: string;
  provider: 'GEMINI' | 'OPENAI' | 'OLLAMA' | 'MOCK' | 'CUSTOM' | string;
  model_name: string;
  has_custom_api_key: boolean;
  masked_api_key?: string | null;
  ollama_base_url: string;
  custom_base_url?: string | null;
  temperature: number;
  pii_masking_enabled: boolean;
  updated_at?: string;
}

export interface WorkspaceAISettingsUpdate {
  provider?: string;
  model_name?: string;
  api_key?: string;
  ollama_base_url?: string;
  custom_base_url?: string;
  temperature?: number;
  pii_masking_enabled?: boolean;
}

export interface TestAIConnectionRequest {
  provider: string;
  model_name?: string;
  api_key?: string;
  ollama_base_url?: string;
  custom_base_url?: string;
}

export interface TestAIConnectionResponse {
  status: 'connected' | 'error';
  latency_ms: number;
  provider: string;
  model_name: string;
  message: string;
  details?: Record<string, any>;
}

export const aiApi = {
  chat: async (
    workspaceId: string,
    message: string,
    conversationId?: string
  ): Promise<ChatResponsePayload> => {
    return apiClient.request<ChatResponsePayload>(`/workspaces/${workspaceId}/ai/chat`, {
      method: 'POST',
      body: JSON.stringify({
        message,
        conversation_id: conversationId,
      }),
    });
  },

  getApprovals: async (
    workspaceId: string,
    status?: 'PENDING' | 'APPROVED' | 'REJECTED' | 'EXECUTED'
  ): Promise<Approval[]> => {
    const qs = status ? `?status=${status}` : '';
    return apiClient.request<Approval[]>(`/workspaces/${workspaceId}/ai/approvals${qs}`);
  },

  approve: async (workspaceId: string, approvalId: string): Promise<any> => {
    return apiClient.request(`/workspaces/${workspaceId}/ai/approvals/${approvalId}/approve`, {
      method: 'POST',
    });
  },

  reject: async (workspaceId: string, approvalId: string, reason?: string): Promise<any> => {
    return apiClient.request(`/workspaces/${workspaceId}/ai/approvals/${approvalId}/reject`, {
      method: 'POST',
      body: JSON.stringify({ reason }),
    });
  },

  getMonitoring: async (workspaceId: string): Promise<AgentTelemetry> => {
    return apiClient.request<AgentTelemetry>(`/workspaces/${workspaceId}/ai/monitoring`);
  },

  getSettings: async (workspaceId: string): Promise<WorkspaceAISettingsResponse> => {
    return apiClient.request<WorkspaceAISettingsResponse>(`/workspaces/${workspaceId}/ai/settings`);
  },

  updateSettings: async (
    workspaceId: string,
    data: WorkspaceAISettingsUpdate
  ): Promise<WorkspaceAISettingsResponse> => {
    return apiClient.request<WorkspaceAISettingsResponse>(`/workspaces/${workspaceId}/ai/settings`, {
      method: 'PUT',
      body: JSON.stringify(data),
    });
  },

  testConnection: async (
    workspaceId: string,
    data: TestAIConnectionRequest
  ): Promise<TestAIConnectionResponse> => {
    return apiClient.request<TestAIConnectionResponse>(
      `/workspaces/${workspaceId}/ai/settings/test-connection`,
      {
        method: 'POST',
        body: JSON.stringify(data),
      }
    );
  },
};

File: backend/app/models/ai.py
from datetime import datetime
from enum import Enum
from typing import List, Optional
from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.base import Base, TimestampMixin, generate_uuid, utc_now


class RiskLevel(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"


class ApprovalStatus(str, Enum):
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    EXECUTED = "EXECUTED"


class AgentRunStatus(str, Enum):
    SUCCESS = "SUCCESS"
    AWAITING_APPROVAL = "AWAITING_APPROVAL"
    ERROR = "ERROR"


class AIConversation(Base, TimestampMixin):
    """Chat thread between a user and the AI Manager Agent within a workspace."""
    __tablename__ = "ai_conversations"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=generate_uuid)
    workspace_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    title: Mapped[str] = mapped_column(String(255), default="AI Manager Consultation", nullable=False)

    # Relationships
    workspace: Mapped["Workspace"] = relationship("Workspace")
    user: Mapped["User"] = relationship("User")
    runs: Mapped[List["AIAgentRun"]] = relationship(
        "AIAgentRun", back_populates="conversation", cascade="all, delete-orphan"
    )


class AIAgentRun(Base):
    """
    Execution trace of a multi-agent orchestration run.
    Records the user query, plan, sub-agent delegations, execution time, and tokens used.
    """
    __tablename__ = "ai_agent_runs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=generate_uuid)
    conversation_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("ai_conversations.id", ondelete="CASCADE"), nullable=False, index=True
    )
    workspace_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    agent_name: Mapped[str] = mapped_column(String(100), default="ManagerAgent", nullable=False)
    user_query: Mapped[str] = mapped_column(Text, nullable=False)
    plan: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    final_response: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    tokens_used: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    cost_estimate: Mapped[float] = mapped_column(Float, default=0.0, nullable=False)
    execution_time_ms: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    status: Mapped[str] = mapped_column(String(50), default=AgentRunStatus.SUCCESS.value, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, nullable=False, index=True
    )

    # Relationships
    conversation: Mapped["AIConversation"] = relationship("AIConversation", back_populates="runs")
    tool_calls: Mapped[List["AIToolCall"]] = relationship(
        "AIToolCall", back_populates="agent_run", cascade="all, delete-orphan"
    )
    approvals: Mapped[List["Approval"]] = relationship(
        "Approval", back_populates="agent_run", cascade="all, delete-orphan"
    )


class AIToolCall(Base):
    """Record of a tool invoked by an agent during execution."""
    __tablename__ = "ai_tool_calls"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=generate_uuid)
    agent_run_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("ai_agent_runs.id", ondelete="CASCADE"), nullable=False, index=True
    )
    tool_name: Mapped[str] = mapped_column(String(100), nullable=False)
    arguments_json: Mapped[str] = mapped_column(Text, nullable=False)
    result_json: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    risk_level: Mapped[str] = mapped_column(String(20), default=RiskLevel.LOW.value, nullable=False)
    is_error: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, nullable=False
    )

    # Relationships
    agent_run: Mapped["AIAgentRun"] = relationship("AIAgentRun", back_populates="tool_calls")


class Approval(Base, TimestampMixin):
    """
    Human Approval Request.
    Medium and High risk actions proposed by AI are staged here.
    AI can NEVER approve its own actions.
    """
    __tablename__ = "ai_approvals"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=generate_uuid)
    workspace_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    agent_run_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("ai_agent_runs.id", ondelete="CASCADE"), nullable=False, index=True
    )
    action_type: Mapped[str] = mapped_column(String(100), nullable=False)
    action_payload: Mapped[str] = mapped_column(Text, nullable=False)
    risk_level: Mapped[str] = mapped_column(String(20), nullable=False)  # MEDIUM or HIGH
    status: Mapped[str] = mapped_column(
        String(50), default=ApprovalStatus.PENDING.value, nullable=False, index=True
    )
    requested_by_agent: Mapped[str] = mapped_column(String(100), nullable=False)
    explanation: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    reviewed_by_id: Mapped[Optional[str]] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    reviewed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    rejection_reason: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    # Relationships
    agent_run: Mapped["AIAgentRun"] = relationship("AIAgentRun", back_populates="approvals")
    reviewed_by: Mapped[Optional["User"]] = relationship("User")


class WorkspaceAISettings(Base, TimestampMixin):
    """
    Configuration for Workspace AI Provider, model selection, BYOK API keys,
    and private Local Ollama endpoints with zero-data-leakage settings.
    """
    __tablename__ = "workspace_ai_settings"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=generate_uuid)
    workspace_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("workspaces.id", ondelete="CASCADE"), nullable=False, unique=True, index=True
    )
    provider: Mapped[str] = mapped_column(String(50), default="MOCK", nullable=False)  # GEMINI, OPENAI, OLLAMA, MOCK
    model_name: Mapped[str] = mapped_column(String(100), default="startupai-mock-v1", nullable=False)
    api_key: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)  # Custom corporate API key
    ollama_base_url: Mapped[str] = mapped_column(String(255), default="http://localhost:11434", nullable=False)
    custom_base_url: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    temperature: Mapped[float] = mapped_column(Float, default=0.2, nullable=False)
    pii_masking_enabled: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    # Relationships
    workspace: Mapped["Workspace"] = relationship("Workspace")

